File: main/algorithms/math/fibonacci/go_weird/fibonacci.go
/*
    type: algorithm
    theme: math
    sub-theme: calculate
    name: fibonacci
    author of code: example is taken from golang.org (official golang site)

*/

package main

import "fmt"

// fib returns a function that returns
// successive Fibonacci numbers.
func fib() func() int {

  a, b := 0, 1

  return func() int {

    a, b = b, a+b

    return a
  }
}

func main() {

  f := fib()

  // Function calls are evaluated left-to-right.
  fmt.Println(f(), f(), f(), f(), f())
}

File: algorithms/math/fibonacci/fibonacci.py
'''
    type: algorithm
    theme: math
    sub-theme: calculate
    name: fibonacci
    author of code: Evgeny @missingdays Bovykin

'''

def fibonacci(n):

    if n == 1:
        return 0

    a,b = 0, 1 #First two numbers

    for i in range(2, n):
        a,b = b, a + b

    return b

if __name__ == "__main__":
    assert fibonacci(1) == 0
    assert fibonacci(2) == 1
    assert fibonacci(3) == 1
    assert fibonacci(4) == 2
    assert fibonacci(5) == 3
    assert fibonacci(6) == 5
    assert fibonacci(7) == 8
    assert fibonacci(8) == 13
    assert fibonacci(9) == 21
    assert fibonacci(10) == 34

    print("Fibonacci python done")